sql/core/src/main/scala/org/apache/spark/sql/json/JsonRDD.scala

```diff
@@ -306,6 +306,10 @@ private[sql] object JsonRDD extends Logging {
           val parsed = mapper.readValue(record, classOf[Object]) match {
             case map: java.util.Map[_, _] => scalafy(map).asInstanceOf[Map[String, Any]] :: Nil
             case list: java.util.List[_] => scalafy(list).asInstanceOf[Seq[Map[String, Any]]]
+            case _ =>
+              sys.error(
+                s"Failed to parse record $record. Please make sure that " +
+                "every record in the dataset is a JSON object or an array of JSON objects.")
```

Comment:
I'd say something like `each line of the file is a valid json object or array` since its not obvious to non-hadoop people that this is how records might be split.